lib/cousins.ml

```diff
@@ -202,20 +235,28 @@ let descendants base cousins_cnt i j =
   let liste2 = if i > 0 then cousins_cnt.(i - 1).(j - 1) else [] in
   descendants_aux base liste1 liste2
 
-let init_cousins_cnt base max_a_l max_d_l p =
+let init_cousins_cnt conf base p =
+  let max_a_l = max_ancestor_level conf base (get_iper p) mal in
+  let max_d_l = max_descendant_level conf base (get_iper p) mdl in
   match (!cousins_t, !cousins_dates_t) with
   | Some t, Some d_t -> (t, d_t)
   | _, _ ->
       let t', d_t' =
-        let max_l1, max_l2 = max_l1_l2_aux max_a_l max_d_l in
-        Printf.sprintf "******** Compute %d × %d table ********\n" max_l1 max_l2
+        Printf.sprintf "******** Compute %d × %d table ********\n" (max_a_l + 3)
+          (max_d_l + max_a_l + 3)
         |> !GWPARAM.syslog `LOG_WARNING;
-        (* TODO test for Sys.max_array_length *)
+        if
+          max_a_l + 3 > Sys.max_array_length
+          || max_d_l + max_a_l + 3 > Sys.max_array_length
+        then failwith "Cousins table too large for system";
         let () = load_ascends_array base in
         let () = load_couples_array base in
-        let cousins_cnt = Array.make_matrix (max_l2 + 1) (max_l2 + 1) [] in
+        (* +3: there may be more descendants for cousins than my own *)
```

Comment:
the +3 looks like a hack
why is it needed?
do the code still works with cousins that have more than +3 descendants level than you?